### ISSUE TITLE
Calling the `llvmint::cttz_iX` is broken

### ISSUE THREAD

**Issue by gnzlbg** [rust-lang/rust#34382]
I am trying to call the `llvmint::cttz_i8(1i8, false)` intrinsic from the [llvmint](https://github.com/huonw/llvmint) crate and am getting the following error:

``` shell
is_zero_undef argument of bit counting intrinsics must be a constant int
 %17 = call i8 @llvm.cttz.i8(i8 %14, i1 zeroext %16)
```

The problem seems to be that rustc is not propagating `false` to LLVM properly.

This was also reported to the llvmint crate: https://github.com/huonw/llvmint/issues/5

**Comment by nagisa** [comment]
If you find `trailing_zeros` method on integers to be not unstable enough, there’s always the `std::intrinsics` module which contains *proper* implementation of [the intrinsic](http://doc.rust-lang.org/stable/core/intrinsics/fn.cttz.html).

**Comment by gnzlbg** [comment]
Is the handling of zero by that intrinsic defined or undefined? The documentation does not say, and one might want to do both (hence why `llvmint::cttz_iX` takes two parameters).

**Comment by nagisa** [comment]
I think the best correct solution you can hope for here is new intrinsics in `std::intrinsics` which could be added provided there’s a strong usecase for zero-is-undef (performance is unlikely to be one).

re zero: answered on the other issue: https://github.com/rust-lang/rust/issues/34381.

**Comment by gnzlbg** [comment]
> I think the best correct solution you can hope for here is new intrinsics in std::intrinsics which could be added provided there’s a strong usecase for zero-is-undef (performance is unlikely to be one).

So this turned into an XY Problem. 

What I actually wanted was to use the BMI 1.0 tzcnt intrinsic without writting any assembly.

Ideally, `std::intrinsics::cttz` would use this if the architecture supports BMI 1.0. LLVM will use that instruction automatically if available when calling through the `llvm.cttz.iX` intrinsic (and will skip the zero check if it can prove that the value cannot be zero), but since there is no architecture flag for BMI 1.0 I don't know how to tell LLVM via rustc that the architecture supports it (and I don't know if the LLVM version used by rustc implements it, but it is new enough that I would guess it does). 

So to work around the lack of a feature flag I ended up adding my own feature flags to my crate and trying to call llvm intrinsics for the particular architectures directly and running into this.

**Comment by eefriedman** [comment]
`-C target-feature=+bmi`

**Comment by gnzlbg** [comment]
> -C target-feature=+bmi

@eefriedman Thanks! Are those documented somewhere (i've tried googling for the supported target-features without any luck)? Is that for BMI 1.0 or BMI 2.0?

**Comment by gnzlbg** [comment]
Is there a way to detect the target-feature's enabled from within a rust program? Something like `#[target_feature = "bmi"]` ?

**Comment by Mark-Simulacrum** [comment]
I suspect that this is no longer an issue judging by the comments, with the proper solution being `target-feature=+bmi`. The original issue is likely impossible to reproduce today since the llvmint crate has not received updates in 2 years. As such, I'm going to close.
